### PR TITLE
[FLINK-7700] Fix RocksDB state merging

### PR DESCRIPTION
**Pull request by aljoscha**
## What is the purpose of the change

See title

## Brief change log

This fixes state merging in the respective classes and also consolidates the different state tests into `StateBackendTestBase` (this helped discover the bug in the first place).

## Verifying this change

See newly added tests that also verify that the backend is empty at the end of the test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes, when merging windows
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no

R: @StefanRRichter could you please review this?

### CODE REVIEW

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java [hunk 1]
```diff
@@ -1978,4 +1979,25 @@ public File getInstanceBasePath() {
...
+
+		for (Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>> column : kvStateInformation.values()) {
+			RocksIterator rocksIterator = db.newIterator(column.f0);
+			rocksIterator.seekToFirst();
+
+			while (rocksIterator.isValid()) {
+				count++;
+				rocksIterator.next();
+			}
+			rocksIterator.close();
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: Could use try-with-resource here.

Comment (aljoscha) [reply flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: changing

**Comment by aljoscha** [comment]
I think it would be good to have an abstraction for that, yes.

**Comment by aljoscha** [comment]
Merged, Thanks for the reviews!

**Pull request closed without merging**